[docs] moodleapp/*: Migrate Moodle App pages (#85)
moodle/devdocs: docs/guides/moodleapp/development/plugins-upgrade-guide.md

  - `isEnabledForUser` is now optional and defaults to `true`.
  - They can implement a new function called `isEnabledForCourse`; this function will receive the `navOptions` and `admOptions` parameters. If it's not defined, it'll default to `true`.
- The function `prefetchPackage` in the `CoreCourseActivityPrefetchHandlerBase` has changed. If you were using this class to implement your own prefetch handler you might need to update its code.
- `CoreInitDelegate` has been deleted. Now the initialisation of the app is done via Angular’s `APP_INITIALIZER`. Please notice that `APP_INITIALIZER` cannot and shouldn’t be used by plugins.
- The function `getAdditionalDownloadableFiles` in question types now needs to return a list of `CoreWSExternalFile`, it no longer accepts a list of strings.
- Files stored to be uploaded later using `CoreFileUploaderProvider` no longer have an `offline` property, now they’re just instances of `FileEntry`.
- `ionViewCanLeave` function has been renamed to `canLeave`.
- The `onchange` method of the `Network` service is now called `onChange`.

## Supporting both Ionic 3 and Ionic 5

> Your plugin should still support Ionic 3 so it works in devices that haven’t updated the app yet. This can be done by checking the value of `appversioncode` sent by the app.Here you can find an example applied to the `choicegroup` plugin: [Choicegroup plugin](https://github.com/dpalou/moodle-mod_choicegroup/blob/ionic5/classes/output/mobile.php#L52).

Reviewer: ```suggestion
Your plugin should still support Ionic 3 so it works on devices that haven't updated the app yet. This can be done by checking the value of `appversioncode` sent by the app. Here you can find an example applied to the `choicegroup` plugin: [Choicegroup plugin](https://github.com/dpalou/moodle-mod_choicegroup/blob/ionic5/classes/output/mobile.php#L52).
```

> As you can see in that repository, the JS and the templates are duplicated in order to have one file to support Ionic 3 and another file to support Ionic 5. In this example, they are called "ionic3" and "latest", but you can structure this as you prefer. You can also have a single file with different HTML depending on the `appversioncode`. That’s up to you.

Reviewer: ```suggestion
As you can see in that repository, the JS and the templates are duplicated in order to have one file to support Ionic 3 and another file to support Ionic 5. In this example, they are called "ionic3" and "latest", but you can structure this as you prefer. You can also have a single file with different HTML depending on the `appversioncode`. That's up to you.
```

## Is there any example I can look at?

> If you used the app’s code as an example to build your plugin, you can do the same. There are also some plugins that have been updated, for example you can see the following PRs on the `choicegroup` plugin:

Reviewer: ```suggestion
If you used the app's code as an example to build your plugin, you can do the same. There are also some plugins that have been updated, for example, you can see the following PRs on the `choicegroup` plugin:
```

- <https://github.com/ndunand/moodle-mod_choicegroup/pull/149>
- <https://github.com/ndunand/moodle-mod_choicegroup/pull/150>

You can also look at the [Moodle App Plugins Development Guide](./plugins-development-guide), it has been updated to reflect how to write plugins for the latest version of the app.

## Before 3.5

Before 3.5, the app was written using Ionic 1 and Moodle plugins could add mobile support by writing an Angular JS/Ionic module, compiling it to a zip, and including it in the plugin.

> Nowadays, you need to install the [Moodle App Additional Features](https://docs.moodle.org//en/Moodle_app_additional_features) plugin to make these plugins compatible with the latest versions of Moodle.

Reviewer: ```suggestion
Nowadays, you need to install the [[Moodle App Additional Features]] plugin to make these plugins compatible with the latest versions of Moodle.
```

Reviewer: It's better to use [[xxx]] placeholder for references to Mediawiki docs because then it will be easier to identify them in the future ;-)
